adr/ADR-57.md

```diff
@@ -0,0 +1,128 @@
+sourcing/mirroring.
+
+```go
+type StreamSource struct {
+    Name                   string `json:"name"`
+    ConsumerName           string `json:"consumer_name,omitempty"`
+    ConsumerDeliverSubject string `json:"consumer_deliver_subject,omitempty"`
+}
+```
+
+### Consumer delivery state reset API
+
+The ordered consumer implementation relies on the consumer's delivery sequence to start at 1 and increment by 1 for
+every delivered message. Gaps are detected by ensuring this delivery sequence increments monotonically. If a gap is
+detected, the consumer delivery state will need to be reset such that the delivery sequence starts at 1 again.
+
+This is a non-issue for the ephemeral ordered push consumer variant as it creates a new consumer starting at the
+expected sequence if a gap is detected. However, the durable consumer must not be deleted and recreated, since that will
+result in losing interest on an Interest stream and subsequently losing messages.
+
+Therefore, the server will provide an API to reset the consumer delivery state. When the server detects a gap, it will
+call this API to reset the consumer delivery state. This ensures the delivery sequence restarts at 1 and any gaps are
+reliably handled.
+
+This reset API, `$JS.API.CONSUMER.RESET.<STREAM>.<CONSUMER>`, will have the following functionality:
+
+- The consumer will be reset, resembling the delivery state of creating a new consumer with `opt_start_seq` set to the
+  specified sequence.
+- The pending and redelivered messages will always be reset.
+- The delivered stream/consumer sequences will always be reset.
+- The ack floor consumer sequence will always be reset.
+- The ack floor stream sequence will be updated depending on the payload. The next message to be delivered is above this
+  new ack floor.
+- An empty payload will reset the consumer's state, but the ack floor stream sequence will remain the same. (This will
+  be used for the durable sourcing consumer after detecting a gap.)
+- A payload of `{"seq":<seq>}` (with `seq>0`) will update the ack floor stream sequence to be one below the provided
+  sequence. The next message to be delivered has a sequence of `msg.seq >= reset.seq`. A zero-sequence is invalid.
+- Resetting a consumer to a specific sequence will only be allowed on specific consumer configurations.
+    - Only allowed on `DeliverPolicy=all,by_start_sequence,by_start_time`.
+    - If `DeliverPolicy=all`, the reset will always be successful and allow to move forward or backward arbitrarily.
+    - If `DeliverPolicy=by_start_sequence,by_start_time`, the reset will only be successful if
+      `reset.seq >= opt_start_seq`, or if `loadNextMsg(reset.seq).start_time >= opt_start_time`. This is a safety
+      measure to prevent the consumer from being reset to a sequence before what was allowed by the consumer
+      configuration.
+- The response to the reset API call will follow standard JS API conventions. Specifically, returning a "consumer
```

Comment:
prefer we make a new response type that have a specific int64 that indicates the new sequence the stream was set to, and then also whatever config/state else you need to communicate like create does.